src/Message/Communicate.cpp

```diff
@@ -68,43 +68,39 @@ Communicate::~Communicate()
 //exclusive:  OOMPI, MPI or Serial
 #ifdef HAVE_OOMPI
 
-Communicate::Communicate(const mpi_comm_type comm_input)
-  : myMPI_destroy_helper(myMPI), d_groupid(0), d_ngroups(1), GroupLeaderComm(nullptr)
+#ifdef HAVE_MPI
+Communicate::Communicate(const mpi3::communicator &in_comm):
+  d_groupid(0), d_ngroups(1), GroupLeaderComm(nullptr)
 {
-  MPI_Comm_dup(comm_input, &myMPI);
+  comm = mpi3::communicator(in_comm);
+  myMPI = &comm;
   myComm = OOMPI_Intra_comm(myMPI);
-  // TODO: mpi3 needs to define comm
   d_mycontext = myComm.Rank();
   d_ncontexts = myComm.Size();
 }
+#endif
 
 
 Communicate::Communicate(const Communicate& in_comm, int nparts)
-  : myMPI_destroy_helper(myMPI)
 {
   std::vector<int> nplist(nparts+1);
-  int p=FairDivideLow(in_comm.rank(), in_comm.size(), nparts, nplist); //group
-  int q=in_comm.rank()-nplist[p]; //rank within a group
-  MPI_Comm_split(in_comm.getMPI(),p,q,&myMPI);
-  myComm=OOMPI_Intra_comm(myMPI);
-  // TODO: mpi3 needs to define comm
+  int p = FairDivideLow(in_comm.rank(), in_comm.size(), nparts, nplist); //group
+  comm = in_comm.comm.split(p, in_comm.rank());
+  myMPI = &comm;
+  myComm = OOMPI_Intra_comm(myMPI);
+
   d_mycontext=myComm.Rank();
   d_ncontexts=myComm.Size();
   d_groupid=p;
   d_ngroups=nparts;
+
   // create a communicator among group leaders.
-  MPI_Group parent_group, leader_group;
-  MPI_Comm_group(in_comm.getMPI(), &parent_group);
-  MPI_Group_incl(parent_group, nparts, nplist.data(), &leader_group);
-  MPI_Comm leader_comm;
-  MPI_Comm_create(in_comm.getMPI(), leader_group, &leader_comm);
+  nplist.pop_back();
+  mpi3::communicator leader_comm = in_comm.comm.subcomm(nplist);
   if(isGroupLeader())
     GroupLeaderComm = new Communicate(leader_comm);
   else
     GroupLeaderComm = nullptr;
-  MPI_Comm_free(&leader_comm);
-  MPI_Group_free(&leader_group);
```

Comment:
Here I free the groups after using them.
The mpi3 wrapper subcomm routine doesn't do it.
@correaa Do you think a fix is needed in subcomm?